pallets/foreign-investments/src/impls/redeem.rs

```diff
@@ -584,3 +584,158 @@ where
 		}
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use rand::{rngs::StdRng, seq::SliceRandom, SeedableRng};
+
+	use super::*;
+
+	#[derive(Clone, Copy, PartialEq, Debug)]
+	enum CurrencyId {
+		Foreign,
+		Pool,
+	}
+
+	const CONVERSION_RATE: u128 = 3; // 300%
+
+	fn to_pool(foreign_amount: u128) -> u128 {
+		foreign_amount * CONVERSION_RATE
+	}
+
+	fn to_foreign(pool_amount: u128) -> u128 {
+		pool_amount / CONVERSION_RATE
+	}
+
+	type RedeemState = super::RedeemState<u128, CurrencyId>;
+	type RedeemTransition = super::RedeemTransition<u128, CurrencyId>;
+
+	impl RedeemState {
+		fn get_done_amount(&self) -> u128 {
+			match *self {
+				Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+					done_amount, ..
+				} => done_amount,
+				Self::RedeemingAndActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+					done_amount,
+					..
+				} => done_amount,
+				Self::SwapIntoForeignDone { done_swap } => done_swap.amount,
+				Self::RedeemingAndSwapIntoForeignDone { done_swap, .. } => done_swap.amount,
+				_ => 0,
+			}
+		}
+
+		fn get_swap_amount(&self) -> u128 {
+			match *self {
+				Self::ActiveSwapIntoForeignCurrency { swap } => swap.amount,
+				Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone { swap, .. } => {
+					swap.amount
+				}
+				Self::RedeemingAndActiveSwapIntoForeignCurrency { swap, .. } => swap.amount,
+				Self::RedeemingAndActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+					swap,
+					..
+				} => swap.amount,
+				_ => 0,
+			}
+		}
```

Comment:
Could make use of `RedeemState::get_active_swap`
```suggestion
		fn get_swap_amount(&self) -> u128 {
			self.get_active_swap().map(|swap| swap.amount).unwrap_or(0)
		}
```

Reply:
Functional and elegant!